### ISSUE TITLE
"Fixed" JSON Paths may occasionally cause troubles

### ISSUE THREAD

**Issue by EmmetZC** [stoplightio/spectral#1910]
> For support questions, please use the [Stoplight Discord Community](https://discord.com/invite/stoplight). This repository's issues are reserved for feature requests and bug reports. If you are unsure if you are experiencing a bug, our Discord is a great place to start.
>
> **Please delete this section, any any sections below that you don't use, before creating the issue.**

**Describe the bug**
JsonPath for `given` is not working properly when `resolved` set to `false`

**To Reproduce**

1. Given this OpenAPI document 
```
{
    "openapi": "3.0.0",
    "info": {
        "version": "0.0.1",
        "title": "Any Title"
    },
    "servers": [
        {
            "url": "https://example.com"
        }
    ],
    "security": [
        {
            "ApiKeyAuth": []
        }
    ],
    "tags": [
        {
            "name": "Orders",
            "description": "no desc"
        }
    ],
    "paths": {
        "/orders": {
            "get": {
                "tags": [
                    "Orders"
                ],
                "summary": "List All Orders (with empty operationId)",
                "operationId": "",
                "parameters": [
                    {
                        "name": "limit",
                        "in": "query",
                        "required": false,
                        "schema": {
                            "example": 5,
                            "type": "integer"
                        }
                    },
                    {
                        "name": "offset",
                        "in": "query",
                        "required": false,
                        "schema": {
                            "example": 10,
                            "type": "integer"
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "Success",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/OrdersListResponse"
                                }
                            }
                        }
                    }
                }
            },
            "post": {
                "tags": [
                    "Orders"
                ],
                "summary": "CreateOrder",
                "operationId": "insert",
                "requestBody": {
                    "required": true,
                    "content": {
                        "application/json": {
                            "schema": {
                                "$ref": "#/components/schemas/OrdersCreateRequest"
                            }
                        }
                    }
                },
                "responses": {
                    "200": {
                        "description": "Success",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/OrdersEntity"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "components": {
        "securitySchemes": {
            "ApiKeyAuth": {
                "type": "apiKey",
                "in": "header",
                "name": "Authorization"
            }
        },
        "schemas": {
            "OrdersCreateRequest": {
                "type": "object",
                "properties": {
                    "data": {
                        "type": "string"
                    }
                }
            },
            "OrdersEntity": {
                "type": "object",
                "properties": {
                    "transaction_id": {
                        "type": "string"
                    },
                    "data": {
                        "type": "string"
                    }
                }
            },
            "OrdersListResponse": {
                "type": "object",
                "properties": {
                    "data": {
                        "type": "array",
                        "items": {
                            "$ref": "#/components/schemas/OrdersEntity"
                        }
                    }
                }
            }
        }
    }
}
```
2. Given this rule
```
  my-operationid:
    description: All OperationIds should not be empty
    given: $.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )].operationId
    severity: error
    resolved: false
    then:
      function: truthy
```
3. Run spectral-cli with custom rule. Obviously, spectral-cli should report one error based on rule `my-operationid`, since `get /orders`'s `operationId` is empty. While in fact there'll be no error reported instead. If you have printed any log in function `truthy`, you'll see no log printed, which means jsonpath of `given` matches nothing.
5. Now remove/comment out `resolved: false` from the rule and run spectral-cli again. Now you see the error, which matches our expectation.

**Expected behavior**
`resolved` should affect the `$ref`s only, the rule `my-operationid` should work properly no matter what is set to `resolved`.

**Environment (remove any that are not applicable):**
 - Library version: spectral-cli@6.1.0, spectral@6.1.0
 - OS: macos 11.6

**Comment by P0lip** [comment]
Hey!
I'm unable to reproduce it using the examples you provided.
![image](https://user-images.githubusercontent.com/9273484/138598633-8748b5d2-38e2-4fb0-a1dd-52b4cba27ca9.png)
The error gets printed as expected.
I might have a hunch, but to confirm that, would you be able to download a binary from here https://github.com/stoplightio/spectral/releases/tag/v6.1.0 and try to lint the very same document using the very same ruleset to see what the output is?

**Comment by EmmetZC** [comment]
Sry for providing incomplete case. I thought this bug could be reproduced with lone rule `my-operationid` and I was wrong.

Check this full ruleset again if you please:
```yaml
rules:
  my-operation-tags-naming:
    description: Tag should named as C variables
    given: $.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )].tags[*]
    severity: error
    resolved: false
    then:
      function: pattern
      functionOptions:
        match: "^_*[a-zA-Z]\\w*$"
  my-operationid:
    description: All OperationIds should not be empty
    given: $.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )].operationId
    severity: error
    resolved: false
    then:
      function: truthy
```

I tried this very same ruleset and the very same document again, with my local `spectral` and the `spectral-macos` downloaded from https://github.com/stoplightio/spectral/releases/tag/v6.1.0. Both of them reproduced the bug.

**Comment by P0lip** [comment]
Thanks for following up!
I know what the issue is - it's not directly connected to the use of `resolved`, but the paths you use.
They're OK, but unfortunately, they trigger a bit of unexpected behavior in the json path engine we use.
I believe I fixed that some time ago, but I'm uncertain whether I released a new version containing the fix.
Either way, if you need a remedy you can do the following
```yaml
rules:
  my-operation-tags-naming:
    description: Tag should named as C variables
    given: $.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )].tags[*]
    severity: error
    resolved: false
    then:
      function: pattern
      functionOptions:
        match: "^_*[a-zA-Z]\\w*$"
  my-operationid:
    description: All OperationIds should not be empty
    given:
      - $.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )].operationId
      - $..someStupidPropertythatWillneverBeMatched
    severity: error
    resolved: false
    then:
      function: truthy
```

I'll post an update later once the issue is fixed in Spectral.

**Comment by EmmetZC** [comment]
Wow, that is an interesting fix. Thx

**Comment by P0lip** [comment]
In case you're interested in understanding why the above change makes it work, here's an explanation.

tl;dr it's due to an optimization we have in the JSON Path engine

Let's take these two paths you provided in the original ruleset:

- `$.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )].tags[*]`
- `$.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )].operationId`

Spectral passes them over to [`nimma`](https://github.com/p0lip/nimma).

Since in this case, the location of the properties can be determined based on JSON Path expression, we generate a "map" of the document we want to traverse.

This is how the map looks like for the object you posted:
```ts
const zones = {
  "paths": {
    "*": {
      "*": {
        "tags": {
          "*": {}
        }
      }
    }
  }
};
```

That map is consumed by [`zonedTraverse`](https://github.com/P0lip/nimma/blob/master/src/runtime/traverse.mjs#L51) function.
The issue is caused by either of these binary expressions
- https://github.com/P0lip/nimma/blob/master/src/runtime/traverse.mjs#L125
- https://github.com/P0lip/nimma/blob/master/src/runtime/traverse.mjs#L113

I don't know which one exactly, because I didn't have the line number stated in the error.

Now, the fix I proposed essentially disables the generation of that "map" of places to traverse, since `$..xyz` is like "traverse everything in the document" so we skip the generation of that map.
